R/teal_slice.R

```diff
@@ -0,0 +1,431 @@
+#' but only `dataname` and `varname` are mandatory, while the others have default values.
+#' Setting any of the other values to NULL means that these parameters will not be modified
+#' (when setting an existing state) or that they will be determined by data (when creating new a new one).
+#' Each of the common elements corresponds to one private field in `FilterState`
+#' where it is stored and from where it is retrieved when calling `FiterState$get_state`.
+#'
+#' A `teal_slice` can also contain any number of additional elements, passed to `...`
+#' as `name:value` pairs. These are collated into a list and stored in the
+#' `private$extras` field.
+#'
+#' All `teal_slice` elements can be passed as arguments to `FilterState` constructors.
+#' A `teal_slice` can be passed to `FilterState$set_state`, which will modify the state.
+#' However, once a `FilterState` is created, only three values can be set with a `teal_slice`:
+#' `selected`, `keep_na` and `keep_inf`.
+#'
+#' Special consideration is given to the `fixed` element.
+#' This is always a logical flag that defaults to FALSE.
+#' In a `FilterState` instantiated with `fixed = TRUE` **none** of the state features can be changed.
+#'
+#' `filter_var` creates a `teal_slice` object, which specifies a filter for a single variable,
+#' passed to and resolved by `FilterState` objects.
+#' `filter_settings` collates multiple `teal_slice` objects into `teal_slices`,
+#' a complete filter specification. This is used by all classes above `FilterState`
+#' as well as `filter_panel_api` wrapper functions.
+#' `teal_slices` also specifies which variables cannot be filtered
+#' and how observations are tallied, which is resolved by `FilterStates`.
+#'
+#' @param dataname `character(1)` name of data set
+#' @param varname `character(1)` name of variable
+#' @param choices vector specifying allowed choices;
+#'                possibly a subset of values in data;
+#'                type and size depends on variable type
+#' @param selected vector specifying selection;
+#'                 type and size depends on variable type
+#' @param keep_na `logical(0-1)` optional logical flag specifying whether to keep missing values
+#' @param keep_inf `logical(0-1)` optional logical flag specifying whether to keep infinite values
+#' @param fixed `logical(1)` logical flag specifying whether to fix this filter state (i.e. forbid setting state)
+#' @param exclude `named list` of `character` vectors where list names match names of data sets
+#'                 and vector elements match variable names in respective data sets;
+#'                 specifies which variables are not allowed to be filtered
+#' @param count_type `character(1)` string specifying how observations are tallied by these filter states
+#' @param show_all `logical(1)` specifying whether NULL elements should also be printed
+#' @param tss `teal_slices`
+#' @param feature `character(1)` name of `teal_slice` element
```

Comment:
I'm not sure about 'feature.' It feels a bit like feature in the 'variable'/'column' sense. That being said I don't really have an alternative.

Reply:
This name is also up for discussion.

Reply:
discussed in another comment